racket/src/expander/syntax/scope.rkt

```diff
@@ -306,9 +324,23 @@
         s)))
 
 (define (scope>? sc1 sc2)
-  ((scope-id sc1) . > . (scope-id sc2)))
+  (let ([id1 (scope-id sc1)]
+        [id2 (scope-id sc2)])
+    (cond [(and (symbol? id1) (symbol? id2))
+           (string>? (symbol->string id1) (symbol->string id2))]
+          [(and (exact-integer? id1) (exact-integer? id2))
+           (> id1 id2)]
+          [else
+           (symbol? id2)])))
 (define (scope<? sc1 sc2)
-  ((scope-id sc1) . < . (scope-id sc2)))
+  (let ([id1 (scope-id sc1)]
+        [id2 (scope-id sc2)])
+    (cond [(and (symbol? id1) (symbol? id2))
+           (string<? (symbol->string id1) (symbol->string id2))]
+          [(and (exact-integer? id1) (exact-integer? id2))
+           (< id1 id2)]
+          [else
+           (symbol? id1)])))
```

Comment:
It's probably be better to make a subtype of `scope` to hold an identity for an interned scope, and leave `id` alone. One benefit is that `scope<?` and `scope>?` can stay simple and reliably fast. Another is that the ordering-by-timestamp property of scopes (which is tied to a binding heuristic, which could be important for avoiding space leaks – although that would be a small corner case) is not tangled with the interned-versus-uninterned distinction.

If the symbol `id` is really better, use `symbol<?` instead of `string->symbol`, and consider defining `scope>?` in terms of `eq?` ad `scope<?` (which may depend on making `new-scope-id!` atomic with a `box-cas!`, which is probably a good idea anyway).

Minor (and may go way with other adjustments): please always use a newline after `cond`.

Reply:
I thought about this, but it seemed harder, and I figured you’d tell me to do it if you thought it was the right thing. :) For a while I was also worried about attaching bindings to interned scopes, since I didn’t fully understand what I was doing, but I don’t think that’s a problem anymore, so I can certainly change this to be a subtype.